src/reducers/index.ts

```diff
@@ -67,18 +69,28 @@ function reduce(state: StoreState, action: Action) {
   if (action.setGamesList) {
     return { ...state, gamesList: action.setGamesList };
   } else if (action.setMatchesList) {
-    let { matchesList, matchIdToMatchState, ...rest } = state;
-    let newMatchIdToMatchState = {};
+    let {
+      matchesList,
+      matchIdToMatchState,
+      currentMatchIndex,
+      ...rest
+    } = state;
+    let newMatchIdToMatchState: MatchIdToMatchState = {};
     action.setMatchesList.forEach(e => {
       if (e.matchId in matchIdToMatchState) {
         newMatchIdToMatchState[e.matchId] = matchIdToMatchState[e.matchId];
-      } else {
-        newMatchIdToMatchState[e.matchId] = {};
       }
     });
+    if (
```

Comment:
You uncovered another bug of mine (thanks again!)

Change above:
checkCondition(
     'currentMatchIndex is in range',
     state.currentMatchIndex >= -1 &&
       state.currentMatchIndex <= state.matchesList.length
   );

to:

checkCondition(
     'currentMatchIndex is in range',
     state.currentMatchIndex >= -1 &&
       state.currentMatchIndex < state.matchesList.length
   );

I don't know what I was thinking :)
Allowing currentMatchIndex === matchesList.length will result in an out-of-bound access...

Extract it into a function:
function isInRange(currentMatchIndex, matchesList) {
  return currentMatchIndex >= -1 &&
       currentMatchIndex < matchesList.length;
}

And reuse that function here and in the invariant check.
If we're not in range, then set currentMatchIndex to -1  (not to ction.setMatchesList.length)